eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Deleg.hs

```diff
@@ -136,31 +164,42 @@ conwayDelegTransition = do
       checkStakeKeyNotAlreadyRegistered stakeCred dsUnified
       pure $ dState {dsUnified = acceptDepositForStakeKey stakeCred dsUnified ppKeyDeposit}
     ConwayUnRegCert stakeCred sMayDeposit -> do
-      checkStakeKeyIsAlreadyRegistered stakeCred dsUnified
-      checkStakeKeyHasZeroBalance stakeCred dsUnified
+      checkStakeKeyIsRegistered stakeCred dsUnified
+      checkStakeKeyHasZeroBalanceReward stakeCred dsUnified
       forM_ sMayDeposit $ checkDepositAgainstPaidDeposit stakeCred dsUnified
       pure $
         dState
-          { dsUnified = UM.domDeleteAll (Set.singleton stakeCred) dsUnified
+          { dsUnified = UM.domDeleteAll (Set.singleton stakeCred) dsUnified -- QUESTION: @WhatisRT: We auto-un-register him as a DRep?
           }
     ConwayDelegCert stakeCred delegatee -> do
-      checkStakeKeyIsAlreadyRegistered stakeCred dsUnified
+      checkStakeKeyIsRegistered stakeCred dsUnified
+      -- QUESTION: @WhatisRT: One cannot delegate if nothing is staked? This seems wrong because `processDelegation` also
```

Comment:
I'm not super familiar with the UMap. In the spec there's the rewards map, and being present in the domain of that map means you've paid a deposit. If the UMap just adds a bunch of fields to the values then it should have the same semantics I guess. I hope that's documented somewhere?

Reply:
> In the spec there's the rewards map, and being present in the domain of that map means you've paid a deposit.

Yes, it is so even in Haskell :+1: 

> If the UMap just adds a bunch of fields to the values then it should have the same semantics I guess. I hope that's documented somewhere?

Yes indeed. And it is documented in the `UMap` module :+1:. The fields added to just the values happen to be
1. The (reward, deposit) pair
2. The stake pool key
3. The drep (key, script, always-noconfidence, or always-abstain)
4. (inconsequential) The set of pointers.

About the `checkStakeKeyIsRegistered`, it results in a predicate failure unless the stake credential is already registered, that is, the stake credential already has made a staking deposit. Should this check be present or not over here? I'm guessing you mean that it _should_ be present, since the stake deposit is expected to be present for the stake credential at this point, and only delegation should be allowed here.